syne_tune/optimizer/schedulers/searchers/searcher.py

```diff
@@ -470,3 +472,113 @@ def clone_from_state(self, state: dict):
     @property
     def debug_log(self):
         return self._debug_log
+
+
+class GridSearcher(BaseSearcher):
+    def __init__(self, config_space, metric, points_to_evaluate=None, **kwargs):
+        super().__init__(config_space, metric, points_to_evaluate)
+        self._hp_ranges = make_hyperparameter_ranges(config_space)
+        self._excl_list = ExclusionList.empty_list(self._hp_ranges)
+        self._config_candidates = self._generate_config_candidates(config_space)
+        # Debug log printing (switched on by default)
+        debug_log = kwargs.get("debug_log", True)
+        if isinstance(debug_log, bool):
+            if debug_log:
+                self._debug_log = DebugLogPrinter()
+            else:
+                self._debug_log = None
+        else:
+            assert isinstance(debug_log, DebugLogPrinter)
+            self._debug_log = debug_log
+
+    def get_config(self, **kwargs):
+        """Sample a new configuration at the grid of config_space
+        This is done without replacement, so previously returned configs are
+        not suggested again.
+
+        Returns
+        -------
+        A new configuration that is valid, or None if no new config can be
+        suggested.
+        """
+        trial_id = kwargs.get("trial_id")
+        if self._debug_log is not None:
+            self._debug_log.start_get_config("grid", trial_id=trial_id)
+        new_config = self._next_initial_config()
+        if new_config is None:
+            while True:
```

Comment:
Jack can say more - but for our purposes it makes it easier to restore the state of searcher with exclusion list; Also a bit easier to deal with initial configs. Let us know if you feel strongly about this.

Reply:
Also, storing the `remain_candidates_on_grid` makes it easier to support the shuffle option.

Reply:
I think it will be very inefficient for large cardinalities and that we should therefore avoid it. You can also store the index in the list of possibilities in addition to the list of excluded candidates and you wont have to do this reject sampling (nor store the list of candidates since it can be regenerated when you re-instantiate your state for instance).

Reply:
I see that `ExclusionList` which is used here is actually using `set` underneath; So adding / removing stuff from there should be ~constant. Looking at the code it appears that as long as GridSearcher is kept in memory getting new candidate should also be ~constant. What would be exact scenario for inefficiency here? Is it reconstruction of searcher from state? Unless we expect users to evaluate millions of candidates or more, overhead for that is unlikely to be noticeable, and is one time for the rest of GridSearcher being in memory.

Reply:
The problem with the current approach is that the code will become very inefficient for large number of candidates both in term of time and space (and you can easily get many candidates since the number of possibilities grows exponentially with the number of hyperparameters). The current implementation stores all sampled candidates and performs reject sampling, so its runtime is O(n) for candidate generation and also O(n) for space whereas it could store only an index of current position in O(1) space and compute next configuration in O(1) time.

Reply:
I've removed the usage of ExclusionList. Instead, using `_remain_candidates_on_grid` (will be an uniq list) to generate next config, store/restore the searcher